include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h

```diff
@@ -0,0 +1,113 @@
+#pragma once
+
+//
+// Copyright rev.ng Labs Srl. See LICENSE.md for details.
+//
+
+#include <map>
+
+#include "llvm/IR/CFG.h"
+#include "llvm/Pass.h"
+
+#include "revng/ADT/GenericGraph.h"
+
+#include "CycleEquivalenceClass.h"
+
+// `struct` used to represent the unique incremental `Index` that is
+// associated with each undirected edge in the `GenericGraph`
+struct EdgeIndex {
+private:
+  size_t Index = 0;
+
+public:
+  EdgeIndex(size_t Index) { this->Index = Index; }
+  size_t getValue() const { return Index; }
+  bool operator<(const EdgeIndex &Rhs) const { return Index < Rhs.getValue(); }
+  bool operator==(const EdgeIndex &Rhs) const {
+    return Index == Rhs.getValue();
+  }
+};
+
+template<class GraphT, class GT = llvm::GraphTraits<GraphT>>
+class CycleEquivalenceAnalysis {
+public:
+  using NodeT = typename GT::NodeRef;
+
+private:
+  struct BlockBidirectionalNode {
+    BlockBidirectionalNode(NodeT Block) : Block(Block) {}
+    NodeT Block;
+    NodeT getBlock() { return Block; }
+    llvm::StringRef getName() const { return Block->getName(); }
+  };
+
+public:
+  using EdgeLabel = EdgeIndex;
+  using BlockNode = BidirectionalNode<BlockBidirectionalNode, EdgeLabel>;
+  using BlockGraph = GenericGraph<BlockNode>;
+  using EdgeDescriptor = revng::detail::EdgeDescriptor<BlockNode *>;
+  using EdgesVectorType = llvm::SmallVector<EdgeDescriptor>;
+  using BracketDescriptor = std::pair<EdgeDescriptor, size_t>;
```

Comment:
What does the `size_t` represent in the pair?

Reply:
It represents the size of the stack of open brackets in the moment the `BracketDescriptor` is created. The pair together represents the equivalence class. I added here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h#L124 a comment explaining better what this field representes.